Cubed sphere grid, projection and mesh generator

danholdaway opened the pull request:
This PR adds the cubed-sphere grid, projections and mesh generator. Included projections are equiangular, equidistance and the fv3 equidistance. It also adds options for rotating the entire grid in the longitudinal direction and for applying a Schmidt stretching, as done in the FV3 stand alone regional (SAR) model. Tests for the three kinds of projections and covering the rotation and stretching options are added. There is still some work to do: namely incorporate the domain decomposition. I'll start on that asap. In the grid class x and y are really just indices in the grid but with the proper rotations. The projection takes in indices (called x, y, tile) and gives back lonlat. It basically means that dx and dy = 1. This is perhaps different to the structured grid class, where there are x and y space grid points, but it's not straightforward to do something like that for the cubed sphere. Perhaps in the future we can investigate using e.g. the quadrilateralized spherical cube mapping, however it is not clear that spacing would be uniform for any of the projections, so it may not provide much benefit. Nevertheless an appropriate mesh can be formed, as seen in the below images.

Cubed-sphere grid:
![fv3](https://user-images.githubusercontent.com/27729500/80750814-be72bd80-8af6-11ea-802b-99a0fdd7b708.png)

2D x and y:
![mesh_csed_xy](https://user-images.githubusercontent.com/27729500/80750831-c6caf880-8af6-11ea-9e82-e576aa06c250.png)

2D lonlat:
![mesh_csed_ll](https://user-images.githubusercontent.com/27729500/80750851-ce8a9d00-8af6-11ea-9a64-4004e50ceb31.png)

FV3-SAR like grid:
![Stretched](https://user-images.githubusercontent.com/27729500/80750491-2f65a580-8af6-11ea-805d-4ae8ae3e45da.png)

danholdaway commented:
@wdeconinck @pmaciel I'm not sure I'm able to assign reviewers on the ecmwf side so perhaps you can assign yourselves?

aerorahul commented:
@danholdaway CubedSphere grid is a structured grid is it not?

danholdaway commented:
@aerorahul it is.

danholdaway commented:
Some failures in the tests due to some strange values for certain compilers, will try to investigate.

aerorahul commented on src/atlas/grid/CubedSphereGrid.h:
```diff
@@ -0,0 +1,81 @@
+class EquiAnglCubedSphereGrid;
+
+/*
+                                             Grid
+                                               |
+                                    +----------+----------+------------------------+
+                                    |                     |                        |
+                             StructuredGrid        UnstructuredGrid           CubedSphere
+                                    |                                              |
+               +--------------------+-----------------------+               +------+------+
+               |                    |                       |               |             |
+          ReducedGrid          GaussianGrid            RegularGrid       EquiDist      EquiAngl
+               |                 |     |                 |     |
+               +--------+--------+     +--------+--------+     +-----+
+                        |                       |                    |
+               ReducedGaussianGrid     RegularGaussianGrid    RegularLonLatGrid
+*/
```
@danholdaway 
Then wouldn't `CubedSphere` be inheriting from `StructuredGrid` as opposed to `Grid`
Or is `StructuredGrid` incapable of defining the `tile` axis?

danholdaway replied:
Yeah, the tile dimension throws in a wrinkle. Think of the image "2D x and y" it would have to represented with a continuous 2D array, resulting in many ghost nodes a very messy grid iterator. Having the tile dimension makes everything a lot easier.

wdeconinck replied:
@aerorahul The `atlas::StructuredGrid` itself is a "one-tile" grid (regional rectangular grids, likely with projections, or global longitude-latitude grids), where points are organized in rows (parallels in y-direction) which are itself not necessarily equidistant. Points on one parallel are equidistant (along x-direction). A cubed sphere grid could not fit in this picture as there are 6 structured tiles.
The best solution would be to create a family of tiled grids, and the cubed sphere, and possibly other grids could fit in this category. But as long as we only have one candidate implementation, I would refrain from already making a TiledGrid base class for this.

aerorahul replied:
@wdeconinck I understand. While `CubedSphere` is a `StructuredGrid`, it is infact a collection of sorts of 6 `StructuredGrid` objects.  I imagine a `TriPolarGrid` will be its own `Grid`.  In essence `StructuredGrid` is really a `RectilinearGrid`.  I guess its just a name.

danholdaway commented:
@wdeconinck the issues with the checks are resolved now.

wdeconinck commented:
Thank you @danholdaway 
I will still did not have time to look in detail, but am looking very much forward in integrating this contribution. In the mean time I would also like to get @MarekWlasak 's input on this.

danholdaway commented:
@MarekWlasak @wdeconinck I've addressed some comments that I received from the Met Office and just pushed them. One of the concerns they flagged was the use of functions like:

`void xy( idx_t i, idx_t j, idx_t t, double xyt[] )`

where variables are passed by value rather than reference. Inside routines like this the array  xyt is modified and it seems to work fine. Indeed if this didn't work the code wouldn't work at all and it clearly does. Further, this kind of approach seems to be used elsewhere in the Atlas source code. Can you advise on using this approach versus a return on the array, which I think becomes a bit more complicated and needs pointers. According to the reviewers at the Met Office functions like this should not be able to return xyt correctly so I wonder if it will cause problems down the road.

MarekWlasak commented:
@wdeconinck @danholdaway - This is what I think is happening- 
'''
double xyt
'''
is copied into the function.  As a result it has a new memory address compared to the input xy field.
xyt can be modified used  inside the routine, but typically will get deallocated once the function has been completed and the local copy goes out of scope.  

However, what we are passing is an array
'''
double xyt[]
'''
Now I think that in this case, while it looks like we are passing by value, what happens is that the array is implicitly treated as a pointer to the first element of the array and being treated as a reference.   This is why it works for you.  ( I wrote my own test case to check this)

However, a more expressive choice could have been
'''
void xy( idx_t i, idx_t j, idx_t t, double (&xyt)[] )
'''

wdeconinck commented:
@danholdaway @MarekWlasak 

It is perfectly OK to pass a C-style array like
```
void function( double xyz[] );
```
It is equivalent to 
```
void function( double *xyz );
```
but has the added user-semantics of being an array, and not a pointer to e.g. a scalar. It is definitely not passed by value.

danholdaway commented:
> @danholdaway @MarekWlasak
> 
> It is perfectly OK to pass a C-style array like
> 
> ```c++
> void function( double xyt[] );
> ```
> 
> It is equivalent to
> 
> ```c++
> void function( double *xyt );
> ```
> 
> but has the added user-semantics of being an array, and not a pointer to e.g. a scalar. It is definitely not passed by value.

Thanks for the clarification @wdeconinck

MarekWlasak commented on src/atlas/projection/detail/ProjectionUtilities.h:
```diff
@@ -0,0 +1,150 @@
+
+  // -----------------------------------------------------------------------------------------------
+
+  static void cartesianToSpherical(double xyz[], double lonlat[], bool right_hand = false) {
```
should this be 
```
const double xyz[], 
```

danholdaway replied:
Added here and elsewhere.

MarekWlasak commented on src/atlas/projection/detail/ProjectionUtilities.h:
```diff
@@ -0,0 +1,150 @@
+
+  static void sphericalToCartesian(double lonlat[], double xyz[], bool right_hand = false,
+                                  bool unit = false) {
+    // Radius
```
```
const double lonlat[]
```

MarekWlasak commented on src/atlas/projection/detail/ProjectionUtilities.h:
```diff
@@ -0,0 +1,150 @@
+
+  //------------------------------------------------------------------------------------------------
+
+  static void rotate3dX(double angle, double xyz[]) {
```
```
const double angle
```

MarekWlasak commented on src/atlas/projection/detail/ProjectionUtilities.h:
```diff
@@ -0,0 +1,150 @@
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
```
Possibly rename ProjectionUtilites.h to SphereProjectionUtilities.h ?

danholdaway replied:
Do we have non-sphere projections? My guess is that Willem will want to do something different with this anyway.

MarekWlasak commented on src/atlas/projection/detail/ProjectionUtilities.h:
```diff
@@ -0,0 +1,150 @@
+
+  //------------------------------------------------------------------------------------------------
+
+  static void rotate3dY(double angle, double xyz[]) {
```
```
const double angle
```

MarekWlasak commented on src/atlas/projection/detail/ProjectionUtilities.h:
```diff
@@ -0,0 +1,150 @@
+
+  //------------------------------------------------------------------------------------------------
+
+  static void rotate3dZ(double angle, double xyz[]) {
```
```
const double angle
```

MarekWlasak commented on src/atlas/projection/detail/ProjectionUtilities.h:
```diff
@@ -0,0 +1,150 @@
+
+  //------------------------------------------------------------------------------------------------
+
+  static void vect_cross(double xyz1[], double xyz2[], double xyzp[]) {
```
```
const double xyz[], const double xyz2[],
```

MarekWlasak commented on src/atlas/projection/detail/ProjectionUtilities.h:
```diff
@@ -0,0 +1,150 @@
+
+  static void mirror_latlon(double lonlat2[], double lonlat3[], double lonlat1[],
+                            double lonlat4[]) {
+    // Given the "mirror" as defined by (lon1, lat1), (lon2, lat2), and center
```
```
const double lonlat2[], const double lonlat3[], const double lonlat1[]
```

MarekWlasak commented on src/atlas/projection/detail/ProjectionUtilities.h:
```diff
@@ -0,0 +1,150 @@
+                            double lonlat4[]) {
+    // Given the "mirror" as defined by (lon1, lat1), (lon2, lat2), and center
+    // of the sphere, compute the mirror image of (lon0, lat0) as  (lon3, lat3)
+
```
// Given the "mirror" as defined by (lon2, lat2), (lon3, lat3), and center
// of the sphere, compute the mirror image of (lon1, lat1) as  (lon4, lat4) ?

danholdaway replied:
Comment deleted

danholdaway commented on doc/example-grids/cubedsphere_equidistant_fv3_sar.yml:
```diff
@@ -0,0 +1,18 @@
+  size : 866
+  lonlat(first) : [359.5,70.4205]
+  lonlat(last)  : [269.5,72.9601]
+  uid : 22b8637a8f74ddba0ebe598c661e18b7
```
I've added the grid stretching to the hash. Good spot!

danholdaway commented on src/atlas/grid/CubedSphereGrid.h:
```diff
@@ -0,0 +1,140 @@
+  Dotted lines: right and top edges of panel
+  >, <, v: direction of increasing index in first dimension
+  *: location of two extra points (ngrid = 6 * (NCube+1) * (NCube+1) + 2)
+
```
I've added an example as you suggest and cleaner up in the mesh generator

danholdaway commented on src/atlas/grid/CubedSphereGrid.h:
```diff
@@ -0,0 +1,140 @@
+
+  // Return the number of tiles
+  inline const int GetNTiles() const { return grid_->GetNTiles(); }
+
```
Done

danholdaway commented on src/atlas/grid/detail/grid/CubedSphere.h:
```diff
@@ -0,0 +1,443 @@
+      return *this;
+    }
+
+    // Given two poisitions in the grid iterator return the distance, which for the cubed-sphere
```
Done

danholdaway commented on src/atlas/grid/detail/grid/CubedSphere.h:
```diff
@@ -0,0 +1,443 @@
+  // Tile specific access to x and y locations
+  // -----------------------------------------
+
+  inline double x123( idx_t i, idx_t j, idx_t t ) const {
```
Removed

danholdaway commented on src/atlas/grid/detail/grid/CubedSphere.h:
```diff
@@ -0,0 +1,443 @@
+    {[this](int i, int j, int t) -> double{return this->y123(i, j, t);},
+     [this](int i, int j, int t) -> double{return this->y456(i, j, t);}
+    };
+
```
Removed.

danholdaway commented on src/atlas/grid/detail/grid/CubedSphere.h:
```diff
@@ -0,0 +1,443 @@
+  inline void xy( idx_t i, idx_t j, idx_t t, double xyt[] ) const {
+    xyt[0] = xtile.at(tileCases_ * t / nTiles_)(i, j, t);
+    xyt[1] = ytile.at(tileCases_ * t / nTiles_)(i, j, t);
+    xyt[2] = static_cast<double>(t);
```
Changed here and elsewhere.

danholdaway commented on src/atlas/grid/detail/grid/CubedSphere.h:
```diff
@@ -0,0 +1,443 @@
+
+  Domain computeDomain() const;
+
+  // Nuber of faces on tile
```
Done

danholdaway commented on src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:
```diff
@@ -0,0 +1,367 @@
+    bool samePoint = ijtproj[0] == ijtnode[0] && ijtproj[1] == ijtnode[1] && ijtproj[2] == ijtnode[2];
+    ATLAS_ASSERT(samePoint, "NOT THE SAME POINT");
+  }
+#endif
```
The unit tests for the grid do not test the mesh. I plan to add testing for the mesh once the Partitioner is added.

danholdaway commented on src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.h:
```diff
@@ -0,0 +1,62 @@
+#include "atlas/array.h"
+#include "atlas/mesh/Connectivity.h"
+#include "atlas/mesh/HybridElements.h"
+#include "atlas/meshgenerator/MeshGenerator.h"
```
Moved

danholdaway commented on src/atlas/meshgenerator/detail/MeshGeneratorFactory.cc:
```diff
@@ -24,6 +25,8 @@ namespace meshgenerator {
     static struct Link {
         Link() {
+            MeshGeneratorBuilder<meshgenerator::CubedSphereMeshGenerator>();
+            MeshGeneratorBuilder<meshgenerator::DelaunayMeshGenerator>();
```
Removed duplicate

danholdaway commented on src/atlas/grid/detail/grid/CubedSphere.h:
```diff
@@ -0,0 +1,443 @@
+  // Nuber of faces on tile
+  idx_t CubeNx_;
+
+  // Nuber of tiles
```
Done

danholdaway commented on src/atlas/grid/detail/grid/CubedSphere.h:
```diff
@@ -0,0 +1,443 @@
+    }
+
+
+    if (ijt[0] == CubeNx_-1 && ijt[1] < CubeNx_-1 && ijt[2] == nTiles_) {  // Final point
```
Bug, fixed.

danholdaway commented on src/atlas/grid/detail/grid/CubedSphere.h:
```diff
@@ -0,0 +1,443 @@
+    return static_cast<double>(xs_[t])  + (y123( i, j, t )-static_cast<double>(ys_[t]));
+  }
+
+  inline double y123( idx_t i, idx_t j, idx_t t ) const {
```
Removed

danholdaway commented on src/atlas/grid/detail/grid/CubedSphere.h:
```diff
@@ -0,0 +1,443 @@
+
+  // Return number of tiles
+  inline const idx_t GetNTiles() const { return nTiles_; }
+
```
Done

danholdaway commented on src/atlas/grid/CubedSphereGrid.h:
```diff
@@ -0,0 +1,140 @@
+  void lonlat2xy( double lonlat[], idx_t ijt[] ) const { grid_->lonlat2xy( lonlat, ijt ); }
+
+  // Return the size of the cubed sphere grid, where CubeNX is the number of grid boxes along the edge of a tile
+  inline const int GetCubeNx() const { return grid_->GetCubeNx(); }
```
Done

danholdaway commented on src/atlas/projection/detail/CubedSphereEquiAnglProjection.cc:
```diff
@@ -0,0 +1,103 @@
+  const auto cubeNx = getCubeNx();
+  auto tile1Lats = getLatArray();
+  auto tile1Lons = getLonArray();
+
```
Is there some example where it would be helpful to not have auto? To my mind this makes the code a cleaner and I'm not sure why it matters that it's hidden. Not opposed to changing I just want to make sure I understand.

danholdaway commented on src/atlas/projection/detail/CubedSphereProjectionBase.cc:
```diff
@@ -0,0 +1,328 @@
+    ATLAS_ASSERT(shiftLon_ >= -90.0, "ShiftLon should be >= -90.0 degrees");
+  };
+
+  // Apply a Shmidt transform
```
Done

danholdaway commented on src/atlas/projection/detail/CubedSphereProjectionBase.cc:
```diff
@@ -0,0 +1,328 @@
+    params.get("ShiftLon", shiftLon_);
+    ATLAS_ASSERT(shiftLon_ <= 90.0,  "ShiftLon should be <= 90.0 degrees");
+    ATLAS_ASSERT(shiftLon_ >= -90.0, "ShiftLon should be >= -90.0 degrees");
+  };
```
Removed here and elsewhere.

danholdaway commented on src/atlas/projection/detail/CubedSphereProjectionBase.cc:
```diff
@@ -0,0 +1,328 @@
+  double lonlat[2];
+  double xyz[3];
+  double angle;
+
```
Removed

danholdaway commented on src/atlas/projection/detail/CubedSphereProjectionBase.h:
```diff
@@ -0,0 +1,98 @@
+    };
+
+    int getCubeNx() const { return cubeNx_; };
+
```
Removed

danholdaway commented:
Thanks for the thorough review @MarekWlasak. I have addressed your comments and marked resolved where I directly took your suggestions and left unresolved where I didn't change anything. Regarding the testing. The projection utilities are tested indirectly. The grid iterator goes through the grid and returns lonlat at the fist and final point. If there was a problem with the projection or the utilities these values would not be correct (match what FV3 produces). These are the values in the yaml files and set manually based on looking at output from FV3. I'll add more testing of the mesh and projection once I add the Partitioner but for now this is as much testing as it given to the other grids. I'm not opposed to comparing with some precomputed mesh file (or something like that) but that would require coordination with Willem since carrying files around might represent a bit of a departure from how they currently test (unless I'm missing something).

wdeconinck commented on src/atlas/projection/detail/CubedSphereProjectionBase.cc:
```diff
@@ -0,0 +1,341 @@
+      lonlat[LON] = lonlat[LON] - 2.0*M_PI;
+    }
+  }
+
```
I wonder how this algorithm compares to a `RotatedSchmidt` already present in Atlas, which takes parameters:
`double stretching_factor`, and extra parameters for rotation, e.g. `PointLonLat north_pole`, `double rotation_angle`. If it is the same, then that would be handy as it already has the inverse projection as well, which seems to be missing.

wdeconinck commented on doc/example-grids/cubedsphere_equidistant_fv3_sar.yml:
```diff
@@ -0,0 +1,18 @@
+DoSchmidt: true
+StretchFac: 3
+TargetLon: -45.5
+TargetLat: 90.0
```
I would have seen rather that this specific rotation/stretching be part of another projection, and be more consistent with the existing RotatedSchmidt projection:
```
projection:
    type: rotated_schmidt
    stretching_factor: 2.0
    north_pole: [-45.5, 90.]
```

danholdaway commented:
No problem at all, thanks for looking at the code @wdeconinck. I'll take a look this week and get back to you with any questions.

wdeconinck commented:
There also seems to be wrong element orientation (inverse node ordering). The convention is:
```
0 --- 3
|     |
1 --- 2
```
or permutations thereof. I can tell because the 2nd and 3rd figure show dark element edges, which indicate the back-side.

danholdaway commented:
Closing as superseded by #66
